osutil/sys/syscall.go

```diff
@@ -107,3 +109,79 @@ func FcntlGetFl(fd int) (int, error) {
 	}
 	return int(flags), nil
 }
+
+func retryRawSyscall(trap, a1, a2, a3 uintptr) (r1, r2 uintptr, err syscall.Errno) {
+	ms := &syscall.Timespec{Nsec: 1000}
+	for i := 0; i < 30; i++ {
+		r1, r2, err = syscall.RawSyscall(trap, a1, a2, a3)
+		if err != syscall.EAGAIN {
+			break
+		}
+		// this could fail, making the loop a little too aggressive
+		syscall.Nanosleep(ms, nil)
+	}
+	return r1, r2, err
+}
+
+type UnrecoverableError struct {
+	Call string
+	Err  error
+}
+
+func (e UnrecoverableError) Error() string {
+	return fmt.Sprintf("%s: %v", e.Call, e.Err)
+}
+
+// RunAsUidGid starts a goroutine, pins it to the OS thread, sets euid and egid,
+// and runs the function; after the function returns, it restores euid and egid.
+//
+// If restoring the original euid and egid fails this function will panic with
+// an UnrecoverableError, and you should _not_ try to recover from it: the
+// runtime itself is going to be in trouble.
+func RunAsUidGid(uid UserID, gid GroupID, f func() error) error {
+	ch := make(chan error, 1)
+	go func() {
+		// from the docs:
+		//   until the goroutine exits or calls UnlockOSThread, it will
+		//   always execute in this thread, and no other goroutine can.
+		// that last bit means it's safe to setuid/setgid in here, as no
+		// other code will run.
+		runtime.LockOSThread()
+
+		// from Go 1.10 on we could just not unlock, which would make
+		// the thread get reaped at goroutine exit. Instead we need to
+		// carefully restore thread state.
+		defer runtime.UnlockOSThread()
+
+		ruid := Getuid()
```

Comment:
While reading I read the code for those. Is there any reason we are not using `getres{u,g}id()`? It uses one less system call.